### PR TITLE
performance: block validation benchmark

### PR DESCRIPTION
**Pull request by algonautshant**
Block validation benchmark
resolves https://github.com/algorand/go-algorand-internal/issues/2259

`$ go test -run none -v -bench BenchmarkBlockValidation  -timeout=500000s -benchtime=10x`

ld: warning: -no_pie is deprecated when targeting new OS versions
goos: darwin
goarch: amd64
pkg: github.com/algorand/go-algorand/ledger
cpu: Intel(R) Core(TM) i7-8559U CPU @ 2.70GHz

BenchmarkBlockValidationJustPayNoNew
Preparing... /1: 100% 1.7s 
Summary 1 blocks and 27026 txns: pay 27026/blk (100%) assets 0/blk (0%) apps 0/blk (0%)
1.1 sec / 1 blks
Preparing... /10: 20% 3.4s 40% 3.4s 60% 3.4s 80% 3.3s 100% 3.4s 
**Summary 10 blocks and 265337 txns: pay 26533/blk (100%) assets 0/blk (0%) apps 0/blk (0%)
11.0 sec / 10 blks**
BenchmarkBlockValidationJustPayNoNew-8   	      10	1099660242 ns/op

BenchmarkBlockValidationJustPay
Preparing... /1: 100% 2.2s 
Summary 1 blocks and 26752 txns: pay 26752/blk (100%) assets 0/blk (0%) apps 0/blk (0%)
1.5 sec / 1 blks
Preparing... /10: 20% 4.3s 40% 4.3s 60% 4.2s 80% 4.4s 100% 4.7s 
**Summary 10 blocks and 262667 txns: pay 26266/blk (100%) assets 0/blk (0%) apps 0/blk (0%)
14.8 sec / 10 blks**
BenchmarkBlockValidationJustPay-8        	      10	1479483790 ns/op

BenchmarkBlockValidationNoNew
Preparing... /1: 100% 3.2s 
Summary 1 blocks and 24581 txns: pay 2930/blk (11%) assets 15453/blk (62%) apps 6198/blk (25%)
2.1 sec / 1 blks
Preparing... /10: 20% 6.0s 40% 5.4s 60% 5.8s 80% 5.4s 100% 6.0s 
**Summary 10 blocks and 239915 txns: pay 3254/blk (13%) assets 14224/blk (59%) apps 6512/blk (27%)
22.7 sec / 10 blks**
BenchmarkBlockValidationNoNew-8          	      10	2266421693 ns/op

BenchmarkBlockValidationMix
Preparing... /1: 100% 2.8s 
Summary 1 blocks and 24773 txns: pay 5905/blk (23%) assets 13339/blk (53%) apps 5529/blk (22%)
2.4 sec / 1 blks
Preparing... /10: 20% 5.4s 40% 6.0s 60% 5.2s 80% 5.6s 100% 6.3s 
**Summary 10 blocks and 242232 txns: pay 5955/blk (24%) assets 12737/blk (52%) apps 5530/blk (22%)**
**23.4 sec / 10 blks**
BenchmarkBlockValidationMix-8            	      10	2343313013 ns/op
PASS
ok  	github.com/algorand/go-algorand/ledger	245.240s

### CODE REVIEW

File: ledger/fullblock_perf_test.go [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+	var blocks []bookkeeping.Block
+	var txPerBlock int
+	var numAss, numPay int
+	fmt.Printf("Preparing transactions and adding the blocks (/%d): ", numBlocks)
+	evalTime := float64(0)
+	addBlockTime := float64(0)
+	s2 := time.Now()
+	s3 := time.Now()
+
+	for i := 0; i < numBlocks; i++ {
```
Comment (algorandskiy) [inline ledger/fullblock_perf_test.go:171]: I suggest to pregenerate numBlocks, then `b.ResetTimer()` and then running another loop adding blocks. Let the bench calculate metrics!

File: ledger/fullblock_perf_test.go [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+	return l0, l1, creator, accts, aIdxs, nil
+}
+
+func BenchmarkBlockValidationMix(b *testing.B) {
+	numAssets := 10000
+	benchmarkBlockValidationMix(b, numAssets)
+}
+
+func BenchmarkBlockValidationPayments(b *testing.B) {
+	benchmarkBlockValidationMix(b, 0)
```
Comment (algorandskiy) [inline ledger/fullblock_perf_test.go:152]: when we pass 0 to setupEnv, it does not create any accounts, except a single creator address, that is not great.
I suggest to have say 50k addresses in a pool for both asset and payment txns, and have a random function for payment receiver: 25% it is a new random address, 75% some existing address.

File: ledger/fullblock_perf_test.go [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+	}
+
+	newBlk := bookkeeping.MakeBlock(blk.BlockHeader)
+	eval, err := l0.StartEvaluator(newBlk.BlockHeader, 5000, 0)
+	if err != nil {
+		return nil, nil, basics.Address{}, nil, nil, err
+	}
+
+	for i, acct := range accts {
+		tx := createAssetTransaction(uint64(i), 1, creator, crypto.Digest{1})
```
Comment (algorandskiy) [inline ledger/fullblock_perf_test.go:111]: maybe we make asset creation as part of regular block validate mix, and not pre-init stage?

File: ledger/fullblock_perf_test.go [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+		stxn := transactions.SignedTxn{Txn: tx, Sig: crypto.Signature{1}}
+		err = eval.Transaction(stxn, transactions.ApplyData{})
+		if err != nil {
+			return nil, nil, basics.Address{}, nil, nil, err
+		}
+
+		// opt in to the asset
+		aIdx := eval.TestingTxnCounter()
+		tx = sendAssetTransaction(uint64(i), 1, acct, 0, crypto.Digest{1}, basics.AssetIndex(aIdx), acct)
```
Comment (algorandskiy) [inline ledger/fullblock_perf_test.go:120]: maybe: for opt in, in order to  use random addresses as receivers, we might modify the protocol and say there is no min balance requirement

File: ledger/fullblock_perf_test.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+	}
+	fmt.Printf("\n%s sec total (eval: %.1fsec  addBlock: %.1fsec)\n", time.Since(s2).String(), evalTime, addBlockTime)
+
+	b.Logf("built %d blocks, each with %d txns: %d assets %d pay", numBlocks, txPerBlock/numBlocks, numAss/numBlocks, numPay/numBlocks)
+
+	// eval + add all the (valid) blocks to the second ledger, measuring it this time
+	vc := verify.GetMockedCache(true)
+	tt := time.Now()
+	b.ResetTimer()
+	for _, blk := range blocks {
```
Comment (algorandskiy) [inline ledger/fullblock_perf_test.go:235]: Shant, I think I miss a point, why do we need two ledgers? to ensure a block we add is valid?

Comment (algonautshant) [reply ledger/fullblock_perf_test.go:253]: exactly. This also simplifies the debugging effort during the implementation of the benchmark. Can and will be removed eventually.

File: ledger/fullblock_perf_test.go (hunk 4 above)
Comment (algonautshant) [reply ledger/fullblock_perf_test.go:106]: This needs to be done for the min balance for account, asset, and app. I think it is still better to first fund the account (to create it) to be more like the real world. 
There will not be any such transaction where the account does not already exist. 
Do you agree?

File: ledger/fullblock_perf_test.go [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+	l0, l1, creator, accts, aIdxs, err :=
+		setupEnv(b.TempDir(), b.Name(), numAssets)
+	require.NoError(b, err)
+
+	numBlocks := b.N
+	cert := agreement.Certificate{}
+	var blocks []bookkeeping.Block
+	var txPerBlock int
+	var numAss, numPay int
+	fmt.Printf("Preparing transactions and adding the blocks (/%d): ", numBlocks)
```
[the remark this answers is not in the document]
Comment (algonautshant) [reply ledger/fullblock_perf_test.go]: b.logf will print at the end. This way it will be more interactive.

**Review by algorandskiy — APPROVED** [review]

**Pull request merged**